opt: fix bug in RejectNullsProject

mgartner opened the pull request:
#### opt: fix bug in RejectNullsProject

This commit fixes a bug in the `RejectNullsProject` rule that pushed
null-rejecting filters below a Project expression for _all_ projected
columns when _any_ of the projected columns had null-rejecting filters
above the projections. In other words, it could incorrectly synthesize a
null-rejecting filter for a projected column, causing incorrect results.

Fixes #100561

Release note (bug fix): A bug has been fixed that could cause queries
with with joins or subqueries to omit rows where column values are NULL
in very rare cases. This bug was present since v20.2.

#### opt: add fast-path to HasNullRejectingFilter

`HasNullRejectingFilter` now returns `false` immediately if the given
column set is empty, rather than unnecessarily extracting null columns
from filter constraints. This is valid because the function only returns
true if the given column set intersects with another column set, and the
empty set never intersects with any other set.

Release note: None

mgartner committed:
opt: fix bug in RejectNullsProject

This commit fixes a bug in the `RejectNullsProject` rule that pushed
null-rejecting filters below a Project expression for _all_ projected
columns when _any_ of the projected columns had null-rejecting filters
above the projections. In other words, it could incorrectly synthesize a
null-rejecting filter for a projected column, causing incorrect results.

Fixes #100561

Release note (bug fix): A bug has been fixed that could cause queries
with with joins or subqueries to omit rows where column values are NULL
in very rare cases. This bug was present since v20.2.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/103611)
<!-- Reviewable:end -->

mgartner commented:
*[`-- commits` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103611#-NVjxpt0AiQJK7PRrbfj):*
It's difficult to define when this bug can occur, and I'm still working on trying to figure out how likely it is to run into. The fact that it's been present for 3 years seems is some evidence that it may be very rare to hit this. It was very difficult to come up with reproductions from the original report, as well. But I haven't discovered a fundamental reason why it would be difficult to hit. Let me know if you have any thoughts @DrewKimball.


<!-- Sent from Reviewable.io -->

mgartner commented:
I'm worried that `RejectNullsGroupBy` also suffers from the same type of bug: https://github.com/cockroachdb/cockroach/blob/6dcfe0fe02ba108e838088a5b425c1ecbc2b041a/pkg/sql/opt/norm/rules/reject_nulls.opt#L117-L151

It appears to create null-rejecting filters for all columns that are requested to be rejected, regardless of it is valid to do so. We should look into this.





<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great find! Thanks for fixing it. 

>I'm worried that `RejectNullsGroupBy` also suffers from the same type of bug

I think you're right from what I see. Do you want to fix that here?

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103611#-:-NVk-oCm7LccpOnLXjUM:b-h62joy)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`-- commits` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103611#-NVjxpt0AiQJK7PRrbfj):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It's difficult to define when this bug can occur, and I'm still working on trying to figure out how likely it is to run into. The fact that it's been present for 3 years seems is some evidence that it may be very rare to hit this. It was very difficult to come up with reproductions from the original report, as well. But I haven't discovered a fundamental reason why it would be difficult to hit. Let me know if you have any thoughts @DrewKimball.
</blockquote></details>

I think it's probably because in most cases null-rejection is requested for all columns in the input of an outer join. Because the nulls in those columns are correlated, it's usually equivalent to null-reject on any of them. In this case, the column that is incorrectly null-rejected is added to the candidate set by a Scan instead of a Join, so it doesn't have the same property. It's probably also made more rare by the fact that only one candidate column is selected for null-rejection, so the incorrect one(s) has to be ordered first by Column ID in the candidate set.


<!-- Sent from Reviewable.io -->

mgartner commented:
> I'm worried that `RejectNullsGroupBy` also suffers from the same type of bug:

I think we're actually safe in `RejectNullsGroupBy` because:

1. We don't request null-rejection for any column if there are multiple aggregate functions: https://github.com/cockroachdb/cockroach/blob/02eb239184c0424f7256e66d391a05380556d7c6/pkg/sql/opt/norm/reject_nulls_funcs.go#L259-L264

2. We only create a single `IS NOT NULL` expression when the rule matches: https://github.com/cockroachdb/cockroach/blob/6dcfe0fe02ba108e838088a5b425c1ecbc2b041a/pkg/sql/opt/norm/rules/reject_nulls.opt#L137-L143

mgartner committed:
opt: add fast-path to HasNullRejectingFilter

`HasNullRejectingFilter` now returns `false` immediately if the given
column set is empty, rather than unnecessarily extracting null columns
from filter constraints. This is valid because the function only returns
true if the given column set intersects with another column set, and the
empty set never intersects with any other set.

Release note: None

mgartner commented:
>  Because the nulls in those columns are correlated, it's usually equivalent to null-reject on any of them.

IIUC, you're saying that if we have table `abcdefgh` on the RHS of a join, and a null-rejecting join filter on `a` (`a = 1`), then it's usually safe to null-reject columns `b, c, d, e, f, g, h`? Doesn't that assume that those columns are usually not-null if `a` is not null?

DrewKimball reviewed (approved):
>IIUC, you're saying that if we have table `abcdefgh` on the RHS of a join, and a null-rejecting join filter on `a` (`a = 1`), then it's usually safe to null-reject columns `b, c, d, e, f, g, h`? Doesn't that assume that those columns are usually not-null if `a` is not null?

Yes, and we also have to have a "null transmitting" scalar operator on both the correct and incorrect columns (like `+` or `>`), and the incorrect one has to be ordered first in the projections list.

> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103611#-:-NVkQbmI0CaDZl7qN4ed:bv0xge7)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

mgartner commented:
> Doesn't that assume that those columns are usually not-null if a is not null?

This seems like a dangerous assumption. It seems quite common that some columns would be null when other are not.

> Yes, and we also have to have a "null transmitting" scalar operator on both the correct and incorrect columns (like + or >), and the incorrect one has to be ordered first in the projections list.

I see. That's definitely makes this less likely to occur.

DrewKimball commented:
>>Doesn't that assume that those columns are usually not-null if a is not null?

>This seems like a dangerous assumption. It seems quite common that some columns would be null when other are not.

True, I only mean that it's a contributing factor, not that it alone makes the bug unlikely. Even if all other conditions are satisfied but the column happens to be non-null, it won't reproduce.

mgartner commented:
I've updated the release note to provide some more information about the bug, but left out the details that I don't think are appropriate for a release note.

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103611#-:-NW8bueLEYIKJ4iTjy2J:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTR!

bors r+

pingxianga commented:
> #### opt: fix bug in RejectNullsProject
> This commit fixes a bug in the `RejectNullsProject` rule that pushed null-rejecting filters below a Project expression for _all_ projected columns when _any_ of the projected columns had null-rejecting filters above the projections. In other words, it could incorrectly synthesize a null-rejecting filter for a projected column, causing incorrect results.
> 
> Fixes #100561
> 
> Release note (bug fix): A bug has been fixed that could cause queries with with joins or subqueries to omit rows where column values are NULL in very rare cases. This bug was present since v20.2.
> 
> #### opt: add fast-path to HasNullRejectingFilter
> `HasNullRejectingFilter` now returns `false` immediately if the given column set is empty, rather than unnecessarily extracting null columns from filter constraints. This is valid because the function only returns true if the given column set intersects with another column set, and the empty set never intersects with any other set.
> 
> Release note: None

Hello sir, I am trying to understand the meaning of this rule, but I cannot understand the meaning of this rule through debugging test cases and consulting information. If you see it, I hope you take the time to answer my questions. Thank you very much, kind sir

pingxianga commented:
> > #### opt: fix bug in RejectNullsProject
> > This commit fixes a bug in the `RejectNullsProject` rule that pushed null-rejecting filters below a Project expression for _all_ projected columns when _any_ of the projected columns had null-rejecting filters above the projections. In other words, it could incorrectly synthesize a null-rejecting filter for a projected column, causing incorrect results.
> > Fixes #100561
> > Release note (bug fix): A bug has been fixed that could cause queries with with joins or subqueries to omit rows where column values are NULL in very rare cases. This bug was present since v20.2.
> > #### opt: add fast-path to HasNullRejectingFilter
> > `HasNullRejectingFilter` now returns `false` immediately if the given column set is empty, rather than unnecessarily extracting null columns from filter constraints. This is valid because the function only returns true if the given column set intersects with another column set, and the empty set never intersects with any other set.
> > Release note: None
> 
> Hello sir, I am trying to understand the meaning of this rule, but I cannot understand the meaning of this rule through debugging test cases and consulting information. If you see it, I hope you take the time to answer my questions. Thank you very much, kind sir

In addition to the comment information I know, I don't see what kind of scenario the real purpose of this rule is for. For example, in general, null value rejection is a necessary condition for converting an outer connection to an inner connection.

pingxianga commented:
> > #### opt: fix bug in RejectNullsProject
> > This commit fixes a bug in the `RejectNullsProject` rule that pushed null-rejecting filters below a Project expression for _all_ projected columns when _any_ of the projected columns had null-rejecting filters above the projections. In other words, it could incorrectly synthesize a null-rejecting filter for a projected column, causing incorrect results.
> > Fixes #100561
> > Release note (bug fix): A bug has been fixed that could cause queries with with joins or subqueries to omit rows where column values are NULL in very rare cases. This bug was present since v20.2.
> > #### opt: add fast-path to HasNullRejectingFilter
> > `HasNullRejectingFilter` now returns `false` immediately if the given column set is empty, rather than unnecessarily extracting null columns from filter constraints. This is valid because the function only returns true if the given column set intersects with another column set, and the empty set never intersects with any other set.
> > Release note: None
> 
> Hello sir, I am trying to understand the meaning of this rule, but I cannot understand the meaning of this rule through debugging test cases and consulting information. If you see it, I hope you take the time to answer my questions. Thank you very much, kind sir

In addition to the comment information I know, I don't see what kind of scenario the real purpose of this rule is for. For example, in general, null value rejection is a necessary condition for converting an outer connection to an inner connection.
